drivers/pwm/pwm_numaker.c

```diff
@@ -0,0 +1,574 @@
+	uint32_t clock_freq;
+	uint32_t cycles_per_sec;
+#ifdef CONFIG_PWM_CAPTURE
+	uint32_t overflows;
+	struct pwm_numaker_capture_data capture[NUMAKER_PWM_CHANNEL_COUNT];
+#endif /* CONFIG_PWM_CAPTURE */
+};
+
+static void pwm_numaker_configure(const struct device *dev)
+{
+	const struct pwm_numaker_config *cfg = dev->config;
+	EPWM_T *epwm = cfg->epwm;
+
+	/* Disable EPWM channel 0~5 before config */
+	EPWM_ForceStop(epwm, 0x3F);
+
+	/* Set EPWM default normal polarity as inverse disabled */
+	epwm->POLCTL &= ~(0x3F << EPWM_POLCTL_PINV0_Pos);
+}
+
+/* PWM api functions */
+static int pwm_numaker_set_cycles(const struct device *dev, uint32_t channel,
+				  uint32_t period_cycles, uint32_t pulse_cycles, pwm_flags_t flags)
+{
+	const struct pwm_numaker_config *cfg = dev->config;
+	struct pwm_numaker_data *data = dev->data;
+	EPWM_T *epwm = cfg->epwm;
+	uint32_t channel_mask = (1 << channel);
+
+	LOG_DBG("===>[%s] Channel=0x%x, CAPIEN=0x%x, CAPIF=0x%x", __func__, channel, epwm->CAPIEN,
+		epwm->CAPIF);
+
+	/* Set EPWM polarity */
+	if (flags & PWM_POLARITY_INVERTED) {
+		epwm->POLCTL |= (1 << (EPWM_POLCTL_PINV0_Pos + channel));
+	} else {
+		epwm->POLCTL &= ~(1 << (EPWM_POLCTL_PINV0_Pos + channel));
+	}
+
+	/* Force disable EPWM channel as while pulse_cycles = 0 */
+	if (period_cycles == 0U) {
+		EPWM_Stop(epwm, 1 << channel);
+		EPWM_ForceStop(epwm, 1 << channel);
+		EPWM_DisableOutput(epwm, 1 << channel);
+		return 0;
+	}
+
+	/* Set EPWM channel & output configuration */
+	EPWM_ConfigOutputChannel(epwm, channel, data->cycles_per_sec / period_cycles,
+				 (100 * pulse_cycles) / period_cycles);
+
+	/* Enable EPWM Output path for EPWM channel */
+	EPWM_EnableOutput(epwm, channel_mask);
+
+	/* Enable Timer for EPWM channel */
+	EPWM_Start(epwm, channel_mask);
+
+	LOG_DBG("===[%s] cycles_per_sec=0x%x, pulse_cycles=0x%x, period_cycles=0x%x", __func__,
+		data->cycles_per_sec, pulse_cycles, period_cycles);
+	LOG_DBG("===[%s] CTL1=0x%x, POEN=0x%x, CNTEN=0x%x", __func__, epwm->CTL1, epwm->POEN,
+		epwm->CNTEN);
+	LOG_DBG("<===[%s] Channel=0x%x, CAPIEN=0x%x, CAPIF=0x%x", __func__, channel, epwm->CAPIEN,
+		epwm->CAPIF);
+
+	return 0;
+}
+
+static int pwm_numaker_get_cycles_per_sec(const struct device *dev, uint32_t channel,
+					  uint64_t *cycles)
+{
+	const struct pwm_numaker_config *cfg = dev->config;
+	struct pwm_numaker_data *data = dev->data;
+
+	ARG_UNUSED(channel);
+
+	data->cycles_per_sec = data->clock_freq / (cfg->prescale + 1U);
+	*cycles = (uint64_t)data->cycles_per_sec;
+
+	return 0;
+}
+
+#ifdef CONFIG_PWM_CAPTURE
+static int pwm_numaker_configure_capture(const struct device *dev, uint32_t channel,
+					 pwm_flags_t flags, pwm_capture_callback_handler_t cb,
+					 void *user_data)
+{
+	struct pwm_numaker_data *data = dev->data;
+	uint32_t pair = channel;
+
+	LOG_DBG("=== Enter %s ...\n", __func__);
+
+	data->capture[pair].callback = cb;
+	data->capture[pair].user_data = user_data;
+
+	if (data->capture[pair].is_busy == true) {
```

Comment:
The `== true` is redundant here since it's a boolean, just use `if (data->capture[pair].is_busy) {` (same on other boolean if conditions)